src/ui/reducers/protocolMessages.ts

```diff
@@ -0,0 +1,74 @@
+import { EventMethods, EventParams } from "@recordreplay/protocol";
+import { createSlice, PayloadAction } from "@reduxjs/toolkit";
+import cloneDeep from "lodash/cloneDeep";
+import { CommandRequest, CommandResponse } from "protocol/socket";
+import { UIState } from "ui/state";
+
+interface Recorded {
+  recordedAt: number;
+}
+
+type _ProtocolEvent<T extends EventMethods> = { method: T; params: EventParams<T> };
+export type ProtocolEvent = _ProtocolEvent<EventMethods>;
+export type RequestSummary = Omit<CommandRequest, "method"> & {
+  class: string;
+  method: string;
+  pending: boolean;
+  errored: boolean;
+  recordedAt: number;
+};
+
+export interface ProtocolMessagesState {
+  events: (ProtocolEvent & Recorded)[];
+  requests: RequestSummary[];
+  responses: (CommandResponse & Recorded)[];
+  errors: (CommandResponse & Recorded)[];
+}
+
+const protocolMessagesSlice = createSlice({
+  initialState: {
+    events: [],
+    requests: [],
+    responses: [],
+    errors: [],
+  } as ProtocolMessagesState,
+  name: "protocolMessages",
+  reducers: {
+    eventReceived(state, action: PayloadAction<ProtocolEvent & Recorded>) {
+      state.events.push(cloneDeep(action.payload));
+    },
+    responseReceived(state, action: PayloadAction<CommandResponse & Recorded>) {
+      state.responses.push(cloneDeep(action.payload));
+      const request = state.requests.find(r => r.id === action.payload.id);
+      if (request) {
+        request.pending = false;
+      }
+    },
+    errorReceived(state, action: PayloadAction<CommandResponse & Recorded>) {
+      state.errors.push(action.payload);
+      const request = state.requests.find(r => r.id === action.payload.id);
+      if (request) {
+        request.pending = false;
+        request.errored = true;
+      }
+    },
+    requestSent(state, action: PayloadAction<CommandRequest & Recorded>) {
+      state.requests.push({
+        ...action.payload,
+        class: action.payload.method.slice(0, action.payload.method.indexOf(".")),
+        method: action.payload.method.slice(action.payload.method.indexOf(".") + 1),
+        pending: !state.responses.some(r => r.id === action.payload.id),
+        errored: state.errors.some(r => r.id === action.payload.id),
+      });
+    },
+  },
+});
+
+export const { errorReceived, eventReceived, requestSent, responseReceived } =
+  protocolMessagesSlice.actions;
+
+export default protocolMessagesSlice;
```

Comment:
:hand: Normally we just export the reducer itself: `export default protocolMessageSlice.reducer`.  Not critical either way, but this matches the "ducks pattern" - default export the reducer, named export the action creators.

Reply:
Ah, perfect, thanks yeah I was trying to figure out the "right" way to do this.